### PR TITLE
Make Copy_if family of APIs use reduce_then_scan algorithm

### PR DESCRIPTION
**Pull request by danhoeflinger**
This PR adds copy_if family to use reduce_then_scan algorithm where it is beneficial.

Moves all algorithm selection decisions to `__parallel_copy_if`, and unifies range API to also use this function.  Adding support for an assignment operator for single work group copy_if.  This allows us to unify the algorithmic selection, and provide performance improvements to the ranges API.

Adds `__parallel_reduce_then_scan_copy` function which will be used by partition and unique in the future to take advantage of shared functionality.

---

This PR is targeted to #1762, to allow for a clean diff, and is a part of the following sequence of PRs meant to be merged in order:

#1769 [MERGED] ~~Relocate __lazy_ctor_storage to utils header~~
#1770 [MERGED] ~~Use __result_and_scratch_storage within scan kernels~~
#1762 Add reduce_then_scan algorithm for transform scan family
#1763 Make Copy_if family of APIs use reduce_then_scan algorithm (This PR)
#1764 Make Partition family of APIs use reduce_then_scan algorithm
#1765 Make Unique family of APIs use reduce_then_scan algorithm

This work is a collaboration between @mmichel11 @adamfidel and @danhoeflinger, and based upon an original prototype by Ted Painter.

### CODE REVIEW

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1028]: I don't like that this change is required, but it mirrors how it is currently handled in the ranges implementation. https://github.com/oneapi-src/oneDPL/blob/aec12a4c55ac5de94753f8e3259b50cd6befb0c3/include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h#L364
Its needed here because we are unifying the ranges API and main API handling of `__parallel_copy_if` to enable the performance improvements to be felt in both cases, and also to remove redundant code.

This is to avoid an issue with making a `zip_view` from a non const lvalue reference to a zip view.  There may be changes we can make to our internal tuple class to resolve this, but I think that lies outside the scope of this PR.

File: include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h
Comment (adamfidel) [inline include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h:387]: In some places, `_Assign` is passed by copy and in others (such as here) it is passed by forwarding ref. Should we try to make these parameters consistent across the board?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (adamfidel) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:853]: I don't see anywhere either in this PR or subsequent PRs where we take advantage of these defaults. It appears that these template parameters are always set. Should we consider removing the defaults?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (mmichel11) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:950]: Similar to the case Adam pointed out with `_Assign`, we take a forwarding reference here with `_Pred` but take it by value in other places.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:853]: Good point, removing defaults.

File: include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h:387]: I think our convention is to just pass by copy elsewhere, I'll make this consistent for both assign and preds in this PR.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (mmichel11) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1067]:
```suggestion
        using _SizeBreakpoints = std::integer_sequence<std::uint16_t, 16, 32, 64, 128, 256, 512, 1024, 2048>;
```

File: include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h]: done

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: done

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: fixed, pass by copy everywhere now to match convention elsewhere in the repo.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: applied.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (mmichel11) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:806]: Since we aren't forwarding this forwarding reference anywhere, would it be better accept `__in_rng` as a const l-value ref: `const _InRng& __in_rng`?

Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:806]: yes, good point. I'll check the other similar helpers to see if there are others to adjust too.

Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: cleaned this up here, and in the previous PR.  I'll scan through the rest of the PRs as well.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:834]: fixed.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:855]: Thanks, fixed, along with `Assign`-> `_Assign`

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: thanks, fixed.

**Review by julianmi — APPROVED** [review]
LGTM. I would hold off merging this until the main patch was merged.

**Review by danhoeflinger — COMMENTED** [review]
This one LGTM, after rebasing minor changes from transform_scan.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (mmichel11) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1034]: Is there a GitHub issue for this? It may be good to create one if there is not.

Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1034]: nope, I'll make one.

**Review by julianmi — DISMISSED** [review]
LGTM. Please wait for the CI before merging this.

**Review by adamfidel — DISMISSED** [review]
This PR looks good to me.

**Review by mmichel11 — APPROVED** [review]
LGTM

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1034]: For others trying to find this issue in the future... #1805

**Pull request merged**
